### ISSUE TITLE
Expanding functionality of styles in basil.js

### ISSUE THREAD

**Issue by trych** [basiljs/basil.js#110]
At the moment working with (character-, paragraph- etc.) styles in basil.js is supported rather rudimentarily. As it is such an essential part of working in InDesign, I think the current implementation could be improved.

I think basil.js should offer the following functionality:
- get the applied style of a given text object
- apply a given style to a given text object or text frame
- create a new style from a given text object

I was giving this some thought and I think to keep it clean the tasks should be split up in three different methods (scope in this case stands for `character|paragraph|object`):
- `b.[scope]Style(name|object)`: remains the getter, gets a style by name or the applied style of a given object.
- `b.apply[Scope]Style(style, textObject|textFrame|object)`: applies a given style to a textObject or textFrame
- `b.create[Scope]Style(textObject|object, name): creates a new style from the given object

Any thoughts?

**Comment by b-g** [comment]
So far the "basil paradigm" is to create alwasy the thing if it isn't there e.g. color(), layer() etc. hence I don't think introducing `b.createXXX` is a good a idea. Hence what about:

* `b.[scope]Style(name|object)`: remains the getter, gets a style by name or the applied style of a given object. **If the style isn't there, basil automatically creates one.**
* `b.apply[Scope]Style(style, textObject|textFrame|object)`: applies a given style to a textObject or textFrame

**Comment by trych** [comment]
The problem in this case is that there are two cases for the getter:

1.) Get an already applied style of a given text object.
2.) Create a new style from the given object (e.g. when the text is red, create a new style with this styling information).

How would the function distinguish between these two cases, if it is combined in the same method? I am totally happy, if we can avoid having three different methods, but this one point I can't quite figure out how to implement.

**Comment by b-g** [comment]
well, I guess:
```
b.[scope]Style(name|object) // get applied style from an object, or get style as defined in doc
b.[scope]Style(name, {props to config new style}) // create new style with props
```
also sidenote, I would change the order of the parameters
`b.apply[Scope]Style(textObject|textFrame|object, style)`

**Comment by trych** [comment]
Ah, now I see the misunderstanding.

What I meant by creating a new style from a given text object was to create a style from an object that *has already all the styling properties applied*. That means I first tint a text red, skew it by 20° etc. and *then* I use the command to create a style from this styling (just like I would do via the InDesign UI: first style my text, then click the new style button).

However, maybe this does not make so much sense in a scripting context. If you style your text with e.g. `b.typo()` one property at a time, just to create a style of it afterwards, you might as well do it via a properties object in the first place. And if you would do the styling by hand beforehand then you could also create the style by hand.

So, long story short, I guess your suggestion makes more sense as there is not much use for my original suggestion.

So, as input for the props object, we just use a regular object and look up the key value pairs?

Something like:

```js
var myStyling = {
skew: 20,
baselineShift: 5,
appliedFont: "Arial"
}

b.characterStyle("myCharStyle", myStyling);
```

Right?

**Comment by fabianmoronzirfas** [comment]
I see the sense of creating a new style from selection. 

A use case would be:  

- User customizes each paragraph/character like he wants to
- User runs script that creates a paragraph- characterstyle per each paragraph/character in selection 

This gives you the advantage to create hundreds of styles without clicking and naming them each on its own and it makes the process repeatable. But we should provide a way to create the names of the styles.

**Comment by fabianmoronzirfas** [comment]
follow up:

It also should allow to see what actually was changed from the base style.

**Comment by fabianmoronzirfas** [comment]
But I also like the idea of the properties object.  

```js
var myStyling = {
skew: 20,
baselineShift: 5,
appliedFont: "Arial"
}
b.characterStyle("myCharStyle", myStyling);
```

**Comment by b-g** [comment]
Hi @trych, yes exactly!

```
var myStyling = {
skew: 20,
baselineShift: 5,
appliedFont: "Arial"
};
b.characterStyle("myCharStyle", myStyling);
```
or you could even write it shorter like
```
b.characterStyle("myCharStyle", {
	skew: 20,
	baselineShift: 5,
	appliedFont: "Arial"
});
```

**Comment by trych** [comment]
@b-g, just started working on this, another question rises right away. You said `b.[scope]Style()` should work like this:

```js
b.style(name|object) // get applied style from an object, or get style as defined in doc
b.style(name, {props to config new style}) // create new style with props
```
So far so good. What should happen however if a name and props are given but it turns out that a style with that name exists? I would suggest that the props object in this case could be used to *update* the existing style.

Same goes for when I have a text object, but still provide a props object as the second argument: the applied style of the text object could be updated with the provided props.

What do you think? Should we implement it like this?

**Comment by b-g** [comment]
@trych Thanks for asking and taking care of this very nitty gritty detail! Sounds good to me! 👍

**Comment by trych** [comment]
This has long been finished via #141.
